std/range/package.d

```diff
@@ -4688,7 +4688,13 @@ if ((isIntegral!(CommonType!(B, E)) || isPointer!(CommonType!(B, E)))
         && isIntegral!S)
 in
 {
-    assert(!(step == 0 && begin != end));
+    import std.conv : unsigned;
+
+    assert(step != 0 || begin == end);
+    if (step > 0)
+        assert(unsigned((end - begin) / step) <= size_t.max);
+    else if (step < 0)
+        assert(unsigned((end - begin) / -step) <= size_t.max);
```

Comment:
And in fact, there is a bug in the code (both in this PR and in `master`):
````
auto i = iota(long.max, long.min, -4);
````
creates a range with only 1 element, and
````
auto i = iota(long.max, long.min, -2);
````
creates an empty range.

But that might be beyond the scope of this PR, since the current code seems to be unable to handle values with that large a range.